### PR TITLE
Defer initialisation of the git repo until the first request is made

### PR DESCRIPTION
**Pull request by knrc**

### CODE REVIEW

File: src/proxy/mod.rs [hunk 1]
```diff
@@ -49,7 +49,7 @@ impl Proxy {
                 );
                 log::info!("Initialising Crate Repository for scope {scope}");
```
Comment (lkatalin) [inline src/proxy/mod.rs:50]: Is it still initialising here? Or should this output be changed?

**Review by lkatalin — COMMENTED** [review]
Thanks, this is a good change.

I am running it locally and trying to understand how it's working with this fix since I am seeing the "initialisation" message before sending any requests to the proxy. I deleted `~/.seedwing-proxy/cache/` to see when it would repopulate. So far the crates are still downloading correctly but I don't see anything under `~/.seedwing-proxy/cache`. Is that expected?

File: src/proxy/mod.rs (hunk 1 above)
Comment (knrc) [reply src/proxy/mod.rs:50]: It's initialising the configuration needed by the git repository, not the local cache of the repository, then displaying it in the logs.  I can change the message if you would like to see something else, just let me know what you would prefer.

**Comment by knrc** [comment]
> I am running it locally and trying to understand how it's working with this fix since I am seeing the "initialisation" message before sending any requests to the proxy. I deleted `~/.seedwing-proxy/cache/` to see when it would repopulate. So far the crates are still downloading correctly but I don't see anything under `~/.seedwing-proxy/cache`. Is that expected?

You likely need to remove move than this, I'm running the following before testing
```rm -rf ~/.seedwing_proxy/ ~/.cargo/registry/*/*localhost* Cargo.lock```

When you start the server you will see the configuration, something like
```
[2023-02-07T00:50:09Z INFO  seedwing_proxy::proxy] ------------------------------------------------------------------------
[2023-02-07T00:50:09Z INFO  seedwing_proxy::proxy] Initialising Crate Repository for scope crates-io
[2023-02-07T00:50:09Z INFO  seedwing_proxy::proxy]     Crate repository       : https://github.com/rust-lang/crates.io-index
[2023-02-07T00:50:09Z INFO  seedwing_proxy::proxy]     local repository cache : /Users/kevin/.seedwing_proxy/cache/crates-io_0.0.0.0_8181
[2023-02-07T00:50:09Z INFO  seedwing_proxy::proxy]     download URL           : http://127.0.0.1:8181/crates-io/api/v1/crates
[2023-02-07T00:50:09Z INFO  seedwing_proxy::proxy]     API URL                : http://127.0.0.1:8181/crates-io
[2023-02-07T00:50:09Z INFO  seedwing_proxy::proxy]     Periodic Update        : 3600
[2023-02-07T00:50:09Z INFO  seedwing_proxy::proxy] ------------------------------------------------------------------------
```

and when you build your project you will see the cache being populated, the checkout of head, the seedwing config changes then then the accesses for the index/crate downloads

```
[2023-02-07T00:57:18Z INFO  seedwing_proxy::repositories::crates::git] Creating cache
[2023-02-07T00:57:18Z INFO  seedwing_proxy::repositories::crates::git] Fetching remote branch: refs/heads/master
[sideband]: Enumerating objects: 232491, done.
[sideband]: Counting objects: 100% (1010/1010), done.
[sideband]: Compressing objects: 100% (545/545), done.
[sideband]: Total 232491 (delta 579), reused 881 (delta 450), pack-reused 231481
[update updated] c0021a020060000020031a020060000080086103..38d1e280c56693f1966ceea2c946e2ef623d66a3 refs/remotes/repository/master
[2023-02-07T00:58:29Z INFO  seedwing_proxy::repositories::crates::git] Tagging 38d1e280c56693f1966ceea2c946e2ef623d66a3
[2023-02-07T00:58:29Z INFO  seedwing_proxy::repositories::crates::git] Checking out head
[2023-02-07T00:59:24Z INFO  seedwing_proxy::repositories::crates::git] Initialising seedwing configuration
[2023-02-07T00:59:27Z INFO  seedwing_proxy::repositories::crates::git] Updating cache
[2023-02-07T00:59:27Z INFO  seedwing_proxy::repositories::crates::git] Fetching remote branch: refs/heads/master
[2023-02-07T00:59:27Z INFO  seedwing_proxy::repositories::crates::git] No updated required
[2023-02-07T00:59:27Z INFO  actix_web::middleware::logger] 127.0.0.1 "GET /crates-io/info/refs?service=git-upload-pack HTTP/1.1" 200 595 "-" "git/2.0 (libgit2 1.5.0)" 129.135502
[2023-02-07T00:59:34Z INFO  actix_web::middleware::logger] 127.0.0.1 "POST /crates-io/git-upload-pack HTTP/1.1" 200 126110614 "-" "git/2.0 (libgit2 1.5.0)" 7.282704
```

**Comment by knrc** [comment]
Note that if you already have an existing seedwing cache for the git repo then you will see something like the following

```
[2023-02-07T01:02:46Z INFO  seedwing_proxy::repositories::crates::git] Updating cache
[2023-02-07T01:02:46Z INFO  seedwing_proxy::repositories::crates::git] Fetching remote branch: refs/heads/master
[sideband]: Enumerating objects: 15, done.
[sideband]: Counting objects: 100% (15/15), done.
[sideband]: Compressing objects: 100% (8/8), done.
[sideband]: Total 10 (delta 6), reused 4 (delta 0), pack-reused 0
[update updated] 38d1e280c56693f1966ceea2c946e2ef623d66a3..1dcd87cc8cb2d55e79cec178a13845ad900f4d68 refs/remotes/repository/master
[2023-02-07T01:02:46Z INFO  seedwing_proxy::repositories::crates::git] Merging 1dcd87cc8cb2d55e79cec178a13845ad900f4d68
[2023-02-07T01:02:48Z INFO  seedwing_proxy::repositories::crates::git] Tagging 1dcd87cc8cb2d55e79cec178a13845ad900f4d68
[2023-02-07T01:02:51Z INFO  actix_web::middleware::logger] 127.0.0.1 "GET /crates-io/info/refs?service=git-upload-pack HTTP/1.1" 200 595 "-" "git/2.0 (libgit2 1.5.0)" 5.103191
```

File: src/proxy/mod.rs (hunk 1 above)
Comment (lkatalin) [reply src/proxy/mod.rs:50]: Ah, interesting. I think we should update the output, but that can be in a different PR as I have a few tweaks I'd like to make to the output in general. I'll open an issue for it.

**Comment by lkatalin** [comment]
@knrc That is a great explanation, it will help document this a bit for posterity as well, thanks.

Interestingly, I wasn't getting the output from `git.rs`, ex. 
`[2023-02-07T16:56:16Z INFO seedwing_proxy::repositories::crates::git] Updating cache`. 
It turned out to be because the `config.toml` file in my example crate was pointing to

```
[source.seedwing]
registry = "file:///<my_dirs>/.seedwing_proxy/cache/crates-io_0.0.0.0_8181/repository/"
```

instead of

```
[source.seedwing]
registry = "http://localhost:8181/crates-io"
```

... yet it still worked fine until I deleted `~/.cargo/registry` entirely. After that, the example crate could not resolve the path to `.../.seedwing-proxy/cache/...` because it didn't exist yet. Now that I've changed the registry, it works fine. :+1:

**Comment by lkatalin** [comment]
Related to #15

**Pull request merged**